src/libraries/System.Linq/tests/CountByTests.cs

```diff
@@ -0,0 +1,116 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Linq.Tests
+{
+    public class CountByTests : EnumerableTests
+    {
+        [Fact]
+        public void CountBy_SourceNull_ThrowsArgumentNullException()
+        {
+            string[] first = null;
+
+            AssertExtensions.Throws<ArgumentNullException>("source", () => first.CountBy(x => x));
+            AssertExtensions.Throws<ArgumentNullException>("source", () => first.CountBy(x => x, new AnagramEqualityComparer()));
+        }
+
+        [Fact]
+        public void CountBy_KeySelectorNull_ThrowsArgumentNullException()
+        {
+            string[] source = { "Bob", "Tim", "Robert", "Chris" };
+            Func<string, string> keySelector = null;
+
+            AssertExtensions.Throws<ArgumentNullException>("keySelector", () => source.CountBy(keySelector));
+            AssertExtensions.Throws<ArgumentNullException>("keySelector", () => source.CountBy(keySelector, new AnagramEqualityComparer()));
+        }
+
+        [Theory]
+        [MemberData(nameof(CountBy_TestData))]
+        public static void CountBy_HasExpectedOutput<TSource, TKey>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? comparer, IEnumerable<KeyValuePair<TKey, int>> expected)
+        {
+            Assert.Equal(expected, source.CountBy(keySelector, comparer));
+        }
+
+        [Theory]
+        [MemberData(nameof(CountBy_TestData))]
+        public static void CountBy_RunOnce_HasExpectedOutput<TSource, TKey>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? comparer, IEnumerable<KeyValuePair<TKey, int>> expected)
+        {
+            Assert.Equal(expected, source.RunOnce().CountBy(keySelector, comparer));
+        }
```

Comment:
Could you also add a couple of tests validating the deferred semantics of the method?

Reply:
Yes, tell me if the new test methods fit the need.